qsimcirq/qsim_simulator.py

```diff
@@ -718,3 +857,28 @@ def _translate_circuit(
             self._translated_circuits.append((circuit, translated_circuit))
 
         return translated_circuit
+
+    def _translate_moments(
+        self,
+        circuit: Any,
+        translator_fn_name: str,
+        qubit_order: cirq.QubitOrderOrList,
+        is_noisy: bool,
+    ):
+        """Translates a circuit by moments.
+
+        Returns:
+            A tuple containing the translated circuit and the gate indices
+            corresponding to moment boundaries.
+        """
+        reindex = [0]
+        for moment in circuit:
+            subc = self._translate_circuit(
+                qsimc.QSimCircuit(cirq.Circuit(moment)),
+                translator_fn_name,
+                qubit_order,
+            )
+            reindex.append(len(subc.channels if is_noisy else subc.gates) + reindex[-1])
+
+        full_circuit = self._translate_circuit(circuit, translator_fn_name, qubit_order)
```

Comment:
The circuit gets translated twice, right? The circuit translation is very slow. Can the code be changed in such a way that the circuit is translated only once (if this is not too much work)?

Reply:
Single translation should be possible - I'll need to add circuit-composition code on the C++ side, but I don't think that should be too complicated. I'll include this in this PR.

Reply:
Done: `compose[_noisy]_circuits` allows Python to merge two C++ circuits together.